<< Compilation stop in: "Built target plugin_charges" >>

<!--
 Before submitting please search open and closed issues at
 https://github.com/openbabel/openbabel

 Open Issues: https://github.com/openbabel/openbabel/issues?q=is%3Aopen+is%3Aissue
 Closed Issues: https://github.com/openbabel/openbabel/issues?q=is%3Aissue+is%3Aclosed

 The GitHub issue tracker is for bug reports and feature requests.
 If you have a question about using Open Babel, please post to openbabel-discuss@lists.sourceforge.net

 Feel free to use the following as a template and remove or add fields as you see fit.
-->

## Environment Information
Open Babel version: 3.1.1
Operating system and version: Ubuntu 22.04.4 LTS, 
GCC: v12


## Expected Behavior
<!--  Describe the intended output or what you expected to see. -->
Compile without errors.
Command used:
`cmake ../openbabel-3.1.1 -DBUILD_GUI=ON -DPYTHON_BINDINGS=ON`


## Actual Behavior
<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->
The compilation process stop as:

```
[ 35%] Linking CXX shared module ../lib/plugin_ops.so
[ 35%] Built target plugin_ops
[ 35%] Linking CXX shared module ../lib/plugin_charges.so
[ 35%] Built target plugin_charges
make: *** [Makefile:146: all] Error 2
```

I searched for compilation issues here. I found issue [1838](https://github.com/openbabel/openbabel/issues/1838), where a (different) compilation problem was solved adding `-DBUILD_SHARED=OFF`, I did that, and the compilation go further, but stopped in another error:

```
[100%] Linking CXX executable ../bin/test_runner
[100%] Built target test_runner
[100%] Linking CXX shared module ../lib/_openbabel.so
/usr/bin/ld: ../src/libopenbabel.a(phmodel.cpp.o): relocation R_X86_64_TPOFF32 against `__tls_guard' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: failed to set dynamic section sizes: bad value
collect2: error: ld returned 1 exit status
make[2]: *** [scripts/CMakeFiles/bindings_python.dir/build.make:102: lib/_openbabel.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:1095: scripts/CMakeFiles/bindings_python.dir/all] Error 2
make: *** [Makefile:146: all] Error 2

```

I met same problem today, and I found the key was in file `obutil.h`. According to error log, `clock` and `CLOCKS_PER_SEC` was not declared in this scope.
```
In file included from /data/openbabel-3.1.1/src/depict/asciipainter.cpp:18:
/data/openbabel-3.1.1/include/openbabel/obutil.h: In member function ‘void OpenBabel::OBStopwatch::Start()’:
/data/openbabel-3.1.1/include/openbabel/obutil.h:65:14: error: ‘clock’ was not declared in this scope
   65 |       start= clock();
      |              ^~~~~
/data/openbabel-3.1.1/include/openbabel/obutil.h:40:1: note: ‘clock’ is defined in header ‘<ctime>’; did you forget to ‘#include <ctime>’?
   39 | #include <math.h>
  +++ |+#include <ctime>
   40 |
/data/openbabel-3.1.1/include/openbabel/obutil.h: In member function ‘double OpenBabel::OBStopwatch::Lap()’:
/data/openbabel-3.1.1/include/openbabel/obutil.h:70:13: error: ‘clock’ was not declared in this scope
   70 |       stop= clock();
      |             ^~~~~
/data/openbabel-3.1.1/include/openbabel/obutil.h:70:13: note: ‘clock’ is defined in header ‘<ctime>’; did you forget to ‘#include <ctime>’?
/data/openbabel-3.1.1/include/openbabel/obutil.h:71:40: error: ‘CLOCKS_PER_SEC’ was not declared in this scope
   71 |       return((stop - start) / (double) CLOCKS_PER_SEC);
      |                                        ^~~~~~~~~~~~~~
/data/openbabel-3.1.1/include/openbabel/obutil.h:71:40: note: the macro ‘CLOCKS_PER_SEC’ had not yet been defined
```
Then I had a look in this file finding them were in  a `if` condition.
```
#if HAVE_CLOCK_T //!<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< HERE
    clock_t start; //!< the start of timing
    clock_t stop;  //!< the current time
#else
    timeval start; //!< the start of timing
    timeval stop;  //!< the current time
#endif

  public:
#if HAVE_CLOCK_T //!<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< HERE

    //! Mark the start of "stopwatch" timing
    void  Start()
    {
      start= clock(); //!<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< HERE
    }
    //! \return The time since calling OBStopwatch::Start() in seconds.
    double Lap()
    {
      stop= clock(); //!<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< HERE
      return((stop - start) / (double) CLOCKS_PER_SEC); //!<<<<<<<<<<<<<<<<<  AND HERE 
    }
#else
    //! Mark the start of "stopwatch" timing
    void Start()
    {
      gettimeofday(&start, nullptr);
    }
    //! \return The time since calling OBStopwatch::Start() in seconds.
    double Lap()
    {
      gettimeofday(&stop, nullptr);
      return((stop.tv_sec - start.tv_sec)
             + (stop.tv_usec - start.tv_usec)/1000000.0);
    }
#endif
```
I am not familiar with C language but when I changed two `HAVE_CLOCK_T` to `false` to forcely skip the first condition, the compilation finished sucessfully. Wish developer would notice this and fix this bug.

--- a/include/openbabel/obutil.h
+++ b/include/openbabel/obutil.h
@@ -35,7 +35,8 @@ GNU General Public License for more details.
 #include <time.h>
 #endif
 #endif
-
+// Include C++ version for clock() and CLOCKS_PER_SEC (required for GCC 12+ with C++17)
+#include <ctime>
 #include <math.h>
 
 #ifndef M_PI

